File: src/dotnet/Core/Application/Services/ExperienceParser.cs
using System.Text.RegularExpressions;

namespace Rag.Candidates.Core.Application.Services;

public sealed class ExperienceParser
{
    private readonly List<Regex> _patterns;

    public ExperienceParser()
    {
        _patterns = new List<Regex>
        {
            new Regex(@"(\d+)\+?\s*(?:years?|yrs?)", RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"at least (\d+)\s*(?:years?|yrs?)", RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"minimum (\d+)\s*(?:years?|yrs?)", RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"min (\d+)\s*(?:years?|yrs?)", RegexOptions.IgnoreCase | RegexOptions.Compiled),
            new Regex(@"(\d+)\s*(?:years?|yrs?)\s*(?:of\s*)?(?:experience|exp)", RegexOptions.IgnoreCase | RegexOptions.Compiled)
        };
    }

    public int? ExtractYears(string queryText)
    {
        foreach (var pattern in _patterns)
        {
            var match = pattern.Match(queryText);
            if (match.Success && match.Groups.Count > 1)
            {
                if (int.TryParse(match.Groups[1].Value, out var years))
                {
                    return years;
                }
            }
        }

        return null;
    }
}

File: src/python/core/application/services/experience_parser.py
import re
from typing import Optional


class ExperienceParser:
    def __init__(self):
        self._patterns = [
            (r'(\d+)\+?\s*(?:years?|yrs?)', r'\1'),
            (r'at least (\d+)\s*(?:years?|yrs?)', r'\1'),
            (r'minimum (\d+)\s*(?:years?|yrs?)', r'\1'),
            (r'min (\d+)\s*(?:years?|yrs?)', r'\1'),
            (r'(\d+)\s*(?:years?|yrs?)\s*(?:of\s*)?(?:experience|exp)', r'\1'),
        ]
    
    def extract_years(self, query_text: str) -> Optional[int]:
        query_lower = query_text.lower()
        
        for pattern, replacement in self._patterns:
            match = re.search(pattern, query_lower)
            if match:
                try:
                    return int(match.group(1))
                except (ValueError, IndexError):
                    continue
        
        return None
